### PR TITLE
livecheck migration: add `brew livecheck` developer command

### PR DESCRIPTION
**Pull request by nandahkrishna**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR integrates `livecheck` into Homebrew/brew as a developer command. It was almost just a copy-paste of files, but there were some style changes and a couple of minor differences as outlined by @samford. I've also brought over the command completions and placed them in the appropriate files (works on `zsh`, should also work on `bash` and `fish` based on my previous trials).

I tested this locally by also incorporating the changes in #8156 (without which we will get an `undefined method strategy` error), and it works as intended (I also untapped homebrew/livecheck).

Things to discuss:
1. What changes do we make to the livecheck code?
2. What tests do we write?
3. (Minor) Should livecheck be a common command (for completions)?

I will also be working on the RuboCops/audits in parallel. Looking forward to your reviews!

### CODE REVIEW

**Comment by nandahkrishna** [comment]
Also, I'm unfamiliar with how the `man` change should be handled (that's why CI seems to failing). I ran `brew man` locally and it made some changes to the man files, if I should include those changes as well, let me know.

**Comment by maxim-belkin** [comment]
> I ran `brew man` locally and it made some changes to the man files, if I should include those changes as well, let me know.

That's exactly what you should do.

**Comment by nandahkrishna** [comment]
Thanks! Just had a look at the `brew license` PR as well and figured it out, should've done that before haha.

**Comment by maxim-belkin** [comment]
I wonder if you could add people who contributed to the files that you add in this PR as co-authors: I think this is the right thing to do with such repo-to-repo migrations. You could do it file-by-file (that'll result in as many commits as there are files -- probably a lot) or once for the single commit.

**Comment by nandahkrishna** [comment]
👍🏼 Totally agree. Maybe it's fine to add everyone as co-authors to the single commit, I think that keeps things cleaner in the PR. However, if you and the others think we should be more granular, I'll do just that.

**Comment by maxim-belkin** [comment]
I don't have a preference -- this is just a way to acknowledge people's contributions on GitHub. Let's do it in a single commit for now.

File: Library/Homebrew/dev-cmd/livecheck.rb
Comment (dawidd6) [inline Library/Homebrew/dev-cmd/livecheck.rb:5]: Could this be:

```suggestion
require "lvecheck_strategy"
```

?

File: Library/Homebrew/dev-cmd/livecheck.rb
Comment (dawidd6) [inline Library/Homebrew/dev-cmd/livecheck.rb:86]: I always wondered what does this line do, any explanations?

Comment (nandahkrishna) [reply Library/Homebrew/dev-cmd/livecheck.rb:86]: There was a `brew livecheck cat` command before, to view a Livecheckable, analogous to `brew cat` to view the Formula source file. We removed that command a while back and this is no longer required (we can just use `brew cat` after the homebrew-core migration), I think. Thanks for pointing it out!

File: Library/Homebrew/dev-cmd/livecheck.rb
Comment (nandahkrishna) [reply Library/Homebrew/dev-cmd/livecheck.rb:5]: Yeah that works, making the change 👍🏼

File: Library/Homebrew/livecheck_strategy.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/livecheck_strategy.rb:3]:
```suggestion
module Homebrew::Livecheck::Strategy
```

File: Library/Homebrew/dev-cmd/livecheck.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/livecheck.rb:75]: Would be good to have this function be as minimal as possible and essentially just parse/setup arguments and pass them to a method in the `Homebrew::Livecheck` module to actually do the main livecheck logic.

File: Library/Homebrew/dev-cmd/livecheck.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/livecheck.rb:108]: I'd say everything here and below should be in the `Homebrew::Livecheck` module.

File: Library/Homebrew/dev-cmd/livecheck.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/livecheck.rb:106]:
```suggestion
    raise UsageError, "No formulae to check." if formulae_to_check.blank?
```

File: Library/Homebrew/livecheck_strategy/apache.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/livecheck_strategy/apache.rb:4]:
```suggestion
  module Homebrew::Livecheck::Strategies::Apache
```

and put it in `livecheck/strategies/apache` using `module_function` to avoid `self.`

File: docs/Manpage.md
Comment (MikeMcQuaid) [inline docs/Manpage.md:888]: Add `HOMEBREW_LIVECHECK_WATCHLIST` to `Library/Homebrew/env_config.rb`

File: Library/Homebrew/dev-cmd/livecheck.rb
Comment (samford) [reply Library/Homebrew/dev-cmd/livecheck.rb:108]: With this in mind, we should probably move the constants that come before `def livecheck_args` (e.g., `WATCHLIST_PATH`, `GITHUB_SPECIAL_CASES`, etc.) into the `Homebrew::Livecheck` module as well, correct? If so, `@livecheck_strategy_names` could also be in `Homebrew::Livecheck`.

File: Library/Homebrew/livecheck_strategy.rb
Comment (samford) [reply Library/Homebrew/livecheck_strategy.rb:3]: If strategies are moving to `livecheck/strategies` (e.g., `livecheck/strategies/apache`) should `livecheck_strategy.rb` be moved into the `livecheck` folder as well? If so, does the file need to be renamed?

Comment (MikeMcQuaid) [reply Library/Homebrew/livecheck_strategy.rb:3]: Yeh, I think so. `livecheck/strategy.rb` or similar may make sense.

File: Library/Homebrew/livecheck_strategy/apache.rb
Comment (samford) [reply Library/Homebrew/livecheck_strategy/apache.rb:4]: We have `Homebrew::Livecheck::Strategy` above and `Homebrew::Livecheck::Strategies::Apache` here. Just to be sure, is the `Strategy` vs. `Strategies` mismatch intentional or should we be doing `Homebrew::Livecheck::Strategy::Apache` here?

Comment (MikeMcQuaid) [reply Library/Homebrew/livecheck_strategy/apache.rb:4]: `Strategy` probably makes more sense to me.

Comment (samford) [reply Library/Homebrew/livecheck_strategy/apache.rb:4]: I agree. It makes sense for `LivecheckStrategy` to become `Homebrew::Livecheck::Strategy` and for `LivecheckStrategy::Apache` to become `Homebrew::Livecheck::Strategy::Apache`.

Last question on this: It seems like we generally align the name of the Ruby file and the related directory (e.g., `version.rb`, `version`). Should we also use a `livecheck/strategy/` directory name, to align with `livecheck/strategy.rb` (as discussed in the other comment)?

Comment (MikeMcQuaid) [reply Library/Homebrew/livecheck_strategy/apache.rb:4]: > Should we also use a `livecheck/strategy/` directory name

Yup 👍🏻

**Comment by nandahkrishna** [comment]
The latest commit includes some refactoring done based on your suggestions @MikeMcQuaid.

I've added a `dev-cmd/livecheck_spec.rb` test similar to `linkage_spec`, but I'm not sure how exactly we'd have to proceed with testing `livecheck`'s core logic. There's a lot of network operations involved and actually testing a Formula might not be ideal as upstream can be unreliable.

How do you think we should approach tests for the strategies and livecheck 'core' code?

**Comment by samford** [comment]
> There's a lot of network operations involved and actually testing a Formula might not be ideal as upstream can be unreliable.

I agree that a lot of livecheck involves network requests and this can make testing more difficult.

How do you feel about stubbing network requests, @MikeMcQuaid? It seems like we could use something like the [`webmock` gem](http://github.com/bblimke/webmock) to avoid external network requests (which can be slow and unreliable) and to control the response for the sake of testing.

File: Library/Homebrew/livecheck/livecheck.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/livecheck/livecheck.rb:148]:
```suggestion
    status_hash["messages"] = messages.presence
```

File: Library/Homebrew/livecheck/strategy/apache.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/livecheck/strategy/apache.rb:11]: This regex (and most of these strategies) is a bit inpeneratrable. You may want to use the multiline regex format, add comments, assign to a constant and explain what `1` and `3` refer to using named match groups instead.

File: Library/Homebrew/livecheck/strategy/git.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/livecheck/strategy/git.rb:11]: Use `system_command` `Utils.popen_*` here instead.

File: Library/Homebrew/livecheck/strategy/git.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/livecheck/strategy/git.rb:16]: Everywhere: prefer `blank?` over `empty?` or `nil?`

File: Library/Homebrew/livecheck/strategy/gnu.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/livecheck/strategy/gnu.rb:14]: Would be nice to have a test that verifies all these formulae still exist in homebrew-core.

File: Library/Homebrew/test/dev-cmd/livecheck_spec.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/test/dev-cmd/livecheck_spec.rb:12]: Thoughts on having this check a single, specified formula?

**Comment by MikeMcQuaid** [comment]
> I've added a `dev-cmd/livecheck_spec.rb` test similar to `linkage_spec`, but I'm not sure how exactly we'd have to proceed with testing `livecheck`'s core logic. There's a lot of network operations involved and actually testing a Formula might not be ideal as upstream can be unreliable.

I think ideally split out the methods based on whether they require a network connection or not. Those that do can have `:needs_network` tests and be really simple. 

> How do you think we should approach tests for the strategies and livecheck 'core' code?

For strategies I'd imagine it'd be a matter of being a bit like the versions tests; pass in some expected correct and incorrect URLs and ensure the strategy gets the expected results.

> How do you feel about stubbing network requests, @MikeMcQuaid? It seems like we could use something like the [`webmock` gem](https://github.com/bblimke/webmock) to avoid external network requests (which can be slow and unreliable) and to control the response for the sake of testing.

We've previously avoided this stubbing because, in my experience, your tests then become tests of the stubbing library more than tests of the actual code. To be clear: I don't expect 100% coverage here and if we can sufficiently extract network logic: that code can have very minimal (and in some cases, no) testing and save the more extensive tests for local logic.

File: Library/Homebrew/dev-cmd/livecheck.rb
Comment (samford) [reply Library/Homebrew/dev-cmd/livecheck.rb:62]: Was `Formula[line]` meant here instead of `Formula[word]`?

File: Library/Homebrew/livecheck/strategy/gnu.rb
Comment (nandahkrishna) [reply Library/Homebrew/livecheck/strategy/gnu.rb:14]: @samford correct me if I'm wrong, but wouldn't the `strategy` feature make `SPECIAL_CASES` obsolete? Wouldn't we just have to use `strategy :page_match` or something to prevent the GNU strategy from being used?

File: Library/Homebrew/test/dev-cmd/livecheck_spec.rb
Comment (nandahkrishna) [reply Library/Homebrew/test/dev-cmd/livecheck_spec.rb:12]: I was thinking of checking the `ruby` Formula, but since upstream is unreliable, crazy idea (I don't know how feasible it is though): how about a 'fake' temporary Formula that checks for Homebrew/brew's latest version or something?

File: Library/Homebrew/livecheck/strategy/gnu.rb
Comment (samford) [reply Library/Homebrew/livecheck/strategy/gnu.rb:14]: Yes, for the `Gnu` and `Sourceforge` `SPECIAL_CASES` we can add/modify the `livecheck` block to use `strategy` instead. I was planning to address this after the Homebrew/brew migration but I can take care of the livecheckable changes if we want to do it as part of this process.

To be clear, the `Gnome` strategy’s `DEV_VERSION_ALLOWLIST` is different. This array constant is used in `#find_versions` to select the proper regex, instead of omitting the URL from matching the strategy.

File: Library/Homebrew/livecheck/livecheck.rb
Comment (nandahkrishna) [reply Library/Homebrew/livecheck/livecheck.rb:141]: Would anywhere mean methods in `livecheck/strategy.rb` or individual strategies as well? Just confirming because we have a `private_class_method :strategies` in `unpack_strategy.rb`.

**Comment by nandahkrishna** [comment]
With the latest push, this PR will only deal with:
* Migrating the `livecheck` command as a `dev-cmd` to Homebrew/brew.
* Tests for the `dev-cmd`.
* Completions for `livecheck`.

I will be creating PRs for the migration of the `livecheck` module and the strategies. This PR depends on those two, and cannot be merged until they have been merged. I'll reference them here after I create them.

File: Library/Homebrew/livecheck/strategy/apache.rb
Comment (nandahkrishna) [reply Library/Homebrew/livecheck/strategy/apache.rb]: I've included this issue in the new PR for strategy migration (in the PR description), so I'll resolve it here.

File: Library/Homebrew/dev-cmd/livecheck.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/livecheck.rb:47]: Is it worth having both? Should this have a header?

File: Library/Homebrew/dev-cmd/livecheck.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/livecheck.rb:54]:
```suggestion
    formulae_to_check = if args.tap
        Tap.fetch(args.tap).formula_names.map { |name| Formula[name] }
    elsif args.installed?
```

File: Library/Homebrew/test/dev-cmd/livecheck_spec.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/test/dev-cmd/livecheck_spec.rb:16]: Let's limit these to a single test (probably this one)

File: Library/Homebrew/dev-cmd/livecheck.rb
Comment (samford) [reply Library/Homebrew/dev-cmd/livecheck.rb:47]: > Is it worth having both?

It looks like [you added `args` to this output](https://github.com/Homebrew/homebrew-livecheck/commit/3d63817fa3974d4ed5594141dfe6d0a2e5588c5e), so I would say it's up to you.

I imagine it would only be useful to have both if/when the `args` parsing isn't working properly. However, we can always add temporary output like this when we're debugging issues, so there may not be much value in the interim time.

If it were up to me, I would probably favor `args` or even removing these two output lines entirely if there aren't other `brew` commands that do the same thing.

**Comment by nandahkrishna** [comment]
The following changes have been made in the latest commit:
* Removed `ARGV.inspect`, retaining only `args`. I couldn't find a command that prints the args but `pr-upload.rb` contains something similar during a dry-run. I decided that it would be okay to just leave the `puts args` there (it's under the `-d` and `-v` switches anyway), unless @MikeMcQuaid feels it isn't necessary.
* Retained only one test (which runs livecheck for a test formula).
* Fixed the `if` style issue.

**Comment by MikeMcQuaid** [comment]
Looks good! Happy for this to merge when CI is 💚. Great work @nandahkrishna!

**Comment by samford** [comment]
This relies on #8254 and 8255 to work, so we probably don't want to merge this PR until after those other two are finished and merged, correct?

**Comment by MikeMcQuaid** [comment]
Agreed.

**Comment by nandahkrishna** [comment]
Incorporating the suggestion and rebasing.

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good to me. We probably want to ensure we delete the existing https://github.com/homebrew/homebrew-livecheck command and library code when we merge this and archive that repository. It'd also be worth adding to `DEPRECATED_OFFICIAL_TAPS` in `Library/Homebrew/official_taps.rb` in this PR, too!

**Comment by nandahkrishna** [comment]
> It'd also be worth adding to `DEPRECATED_OFFICIAL_TAPS` in `Library/Homebrew/official_taps.rb` in this PR, too!
Added this 👍🏼

Will rebase and push after #8544 is merged.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
> We probably want to ensure we delete the existing https://github.com/homebrew/homebrew-livecheck command and library code when we merge this

@nandahkrishna Could you open a PR for this, too? I'll merge it when this is merged.

**Comment by nandahkrishna** [comment]
Sure Mike 👍🏼

**Comment by MikeMcQuaid** [comment]
Thanks again @nandahkrishna!

**Pull request merged**
